### PR TITLE
Add BFT-related storage paths (primary and worker)

### PR DESCRIPTION
**Pull request by niklaslong**
Also removes the obsolete operator paths, addresses some clippy lints and updates the lock file.

### CODE REVIEW

**Review by joske — APPROVED** [review]

File: storage/src/lib.rs [hunk 1]
```diff
@@ -36,23 +36,40 @@ pub fn aleo_dir() -> PathBuf {
     path
 }
 
+///
+/// Returns the workspace path.
+///
+pub fn workspace_dir() -> String {
```
Comment (ljedrz) [inline storage/src/lib.rs:42]: maybe we should return the `Path` or `PathBuf` here instead, for greater versatility :thinking:?

Comment (ljedrz) [reply storage/src/lib.rs:42]: likely the latter, as `output.stdout()` returns a `Vec<u8>`

File: storage/src/lib.rs [hunk 2]
```diff
@@ -36,23 +36,40 @@ pub fn aleo_dir() -> PathBuf {
...
+        .arg("--message-format=plain")
+        .output()
+        .unwrap()
+        .stdout;
+    let cargo_path = Path::new(std::str::from_utf8(&output).unwrap().trim());
+    cargo_path.parent().unwrap().display().to_string()
+}
+
+// Returns the base path for the storage files.
+fn base_storage_path(dev: Option<u16>) -> PathBuf {
```
Comment (ljedrz) [inline storage/src/lib.rs:55]: this might be useful on its own, so I'd `pub` it

File: storage/src/lib.rs [hunk 3]
```diff
@@ -71,71 +88,81 @@ pub fn aleo_ledger_dir(network: u16, dev: Option<u16>) -> PathBuf {
 }
 
 ///
-/// Returns the directory for accessing the operator files from Aleo storage.
+/// Returns the directory for accessing the prover files from Aleo storage.
 ///
-/// In production mode, the expected directory path is `~/.aleo/storage/operator-{network}`.
-/// In development mode, the expected directory path is `/path/to/repo/.operator-{network}-{id}`.
+/// In production mode, the expected directory path is `~/.aleo/storage/prover-{network}`.
+/// In development mode, the expected directory path is `/path/to/repo/.prover-{network}-{id}`.
```
Comment (ljedrz) [inline storage/src/lib.rs:94]: maybe we can move both the ledger and the BFT-related storage to a new `../repo/.storage` path :thinking:?

File: storage/src/lib.rs [hunk 4]
```diff
@@ -71,71 +88,81 @@ pub fn aleo_ledger_dir(network: u16, dev: Option<u16>) -> PathBuf {
...
 }
 
 ///
-/// Returns the directory for accessing the prover files from Aleo storage.
+/// Returns the path for the primary-related BFT files.
 ///
-/// In production mode, the expected directory path is `~/.aleo/storage/prover-{network}`.
-/// In development mode, the expected directory path is `/path/to/repo/.prover-{network}-{id}`.
+/// In production mode, the expected directory path is `~/.aleo/storage/bft-{network}/primary`.
+/// In development mode, the expected directory path is `path/to/repo/.bft-storage-{network}/primary-{id}`.
```
Comment (ljedrz) [inline storage/src/lib.rs:119]: maybe we can move both the ledger and the BFT-related storage to a new `../repo/.storage` path :thinking:?

File: storage/src/lib.rs [hunk 5]
```diff
@@ -36,23 +36,40 @@ pub fn aleo_dir() -> PathBuf {
...
+}
+
+// Returns the base path for the storage files.
+fn base_storage_path(dev: Option<u16>) -> PathBuf {
+    // Retrieve the starting directory.
+    match dev.is_some() {
+        // In development mode, the ledger is stored in the root directory of the repository.
+        true => workspace_dir().into(),
+        // In production mode, the ledger is stored in the `~/.aleo/` directory.
+        false => aleo_dir(),
```
Comment (ljedrz) [inline storage/src/lib.rs:61]: maybe here as well

File: storage/src/lib.rs [hunk 6]
```diff
@@ -22,120 +22,192 @@ const ALEO_DIRECTORY: &str = ".aleo";
...
+    path
+}
+
+///
+/// Returns the base path for the storage files.
+///
+/// The expected paths are:
+///
+/// - Production: `~/.aleo/storage`.
+/// - Development: `../.aleo/storage`.
```
Comment (ljedrz) [inline storage/src/lib.rs:85]:
```suggestion
/// - Development: `<snarkOS>/.aleo/storage`.
```

File: storage/src/lib.rs [hunk 7]
```diff
@@ -22,120 +22,192 @@ const ALEO_DIRECTORY: &str = ".aleo";
...
-        // In development mode, the operator is stored in the repository root directory.
-        true => match std::env::current_dir() {
-            Ok(current_dir) => current_dir,
-            _ => PathBuf::from(env!("CARGO_MANIFEST_DIR")),
-        },
-        // In production mode, the operator is stored in the `~/.aleo/` directory.
-        false => aleo_dir(),
-    };
+/// - Production: `~/.aleo/storage/prover-{network}`.
+/// - Development: `../.aleo/storage/prover-{network}-{id}`.
```
Comment (ljedrz) [inline storage/src/lib.rs:131]:
```suggestion
/// - Development: `<snarkOS>/.aleo/storage/prover-{network}-{dev_id}`.
```

File: storage/src/lib.rs [hunk 8]
```diff
@@ -22,120 +22,192 @@ const ALEO_DIRECTORY: &str = ".aleo";
...
+}
 
-    // Construct the path to the operator in storage.
+///
+/// Returns the path for the primary-related BFT files.
+///
+/// The expected paths are:
+///
+/// - Production: `~/.aleo/storage/bft-{network}/primary`.
+/// - Development: `../.aleo/storage/bft-{network}/primary-{id}`.
```
Comment (ljedrz) [inline storage/src/lib.rs:160]:
```suggestion
/// - Development: `<snarkOS>/.aleo/storage/bft-{network}/primary-{dev_id}`.
```

File: storage/src/lib.rs (hunk 7 above)
Comment (ljedrz) [reply storage/src/lib.rs:131]: also, this one should go into the `bft` subfolder, like the worker

File: storage/src/lib.rs (hunk 8 above)
Comment (ljedrz) [reply storage/src/lib.rs:160]: also, I guess we should follow the ledger and prover logic and do `<snarkOS>/.aleo/storage/bft-{network}-{dev_id}/primary` instead, no?

File: storage/src/lib.rs [hunk 9]
```diff
@@ -22,120 +22,192 @@ const ALEO_DIRECTORY: &str = ".aleo";
...
-/// Returns the directory for accessing the prover files from Aleo storage.
+/// Returns the path for the worker-related BFT files.
 ///
-/// In production mode, the expected directory path is `~/.aleo/storage/prover-{network}`.
-/// In development mode, the expected directory path is `/path/to/repo/.prover-{network}-{id}`.
+/// The expected paths are:
 ///
-pub fn aleo_prover_dir(network: u16, dev: Option<u16>) -> PathBuf {
+/// - Production: `~/.aleo/storage/bft-{network}/worker-{worker-id}`.
+/// - Development: `../.aleo/storage/bft-{network}/worker-{primary_id}-{worker_id}`.
```
Comment (ljedrz) [inline storage/src/lib.rs:189]: same as with the primary

File: storage/src/lib.rs (hunk 8 above)
Comment (niklaslong) [reply storage/src/lib.rs:160]: Good catch!

**Review by ljedrz — APPROVED** [review]

**Comment by howardwu** [comment]
@niklaslong there are some merge conflicts here to resolve, with the latest changes.

File: storage/src/lib.rs [hunk 10]
```diff
@@ -22,120 +22,194 @@ const ALEO_DIRECTORY: &str = ".aleo";
...
-    };
+    let mut path = aleo_base_storage_dir(dev);
 
     // Construct the path to the ledger in storage.
     match dev {
         // In development mode, the ledger files are stored in a hidden folder.
         Some(id) => {
-            path.push(format!(".ledger-{}-{}", network, id));
-            path
+            path.push(format!("ledger-{}-{}", network, id));
```
[the remark this answers is not in the document]
Comment (joske) [reply storage/src/lib.rs:114]: Never mind, understood the change.

**Comment by niklaslong** [comment]
Yep, it has been a while since this was opened, I'll update the branch asap 👍

**Comment by niklaslong** [comment]
Rebased!

**Review by ljedrz — CHANGES_REQUESTED** [review]
Actually, we will need to adjust this setup to account for binary runs; we should probably move the committee files from the workspace dir to the main aleo folder

**Pull request closed without merging**